File: iosapp/iosapp/Views/Components/ButtonView.swift
//
//  Button.swift
//  iosapp
//
//  Created by Max Polkovnik on 17/11/2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import SwiftUI
import shared

struct ButtonView: View {
    private let state: ButtonState
    
    init(_ state: ButtonState) {
        self.state = state
    }
    
    var body: some View {
        Button(
            action: state.onClick,
            label: {
                Text(state.text)
                    .frame(minWidth: 0, maxWidth: .infinity)
                    .padding()
                    .foregroundColor(Color(state.textColor.toUIColor()))
                    .overlay(
                        RoundedRectangle(cornerRadius: CGFloat(state.cornerRadius))
                            .stroke(Color(state.outlineColor.toUIColor()), lineWidth: 1)
                )
            }
        )
            .background(Color(state.backgroundColor.toUIColor())) // If you have this
            .cornerRadius(CGFloat(state.cornerRadius))
            .disabled(!state.isEnabled)
    }
}

struct ButtonView_Previews: PreviewProvider {
    static var previews: some View {
        VStack {
            ButtonView(ButtonState.companion.Blue(text: "Hello", onClick: {} ))
            ButtonView(ButtonState.companion.BlueDisable(text: "Hello"))
            ButtonView(ButtonState.companion.BlueOutline(text: "Hello", onClick: {} ))
            ButtonView(ButtonState.companion.BlueOutlineDisabled(text: "Hello"))
            ButtonView(ButtonState.companion.BlueTextOnly(text: "Hello", onClick: {} ))
            ButtonView(ButtonState.companion.BlueTextOnlyDisabled(text: "Hello"))
        }
    }
}

File: components/src/androidMain/kotlin/polkovnik/max/components/ButtonView.kt
package polkovnik.max.components

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp

@Composable
fun ButtonView(state: ButtonState, modifier: Modifier = Modifier) {
    Button(
        onClick = { state.onClick() },
        shape = RoundedCornerShape(state.cornerRadius.dp),
        border = BorderStroke(
            width = 1.dp,
            color = state.outlineColor.toComposeColor()
                .takeIf { state.isOutlineVisible }
                ?: Color.Transparent
        ),
        colors = ButtonDefaults.buttonColors(
            containerColor = state.backgroundColor.toComposeColor(),
            disabledContainerColor = state.backgroundColor.toComposeColor(),
        ),
        enabled = state.isEnabled,
        modifier = modifier,
    ) {
        Text(
            text = state.text,
            color = state.textColor.toComposeColor()
        )
    }
}

@Preview
@Composable
fun ButtonViewPreview() {
    Column {
        ButtonView(ButtonState.Blue("Press me!") {})
        ButtonView(ButtonState.BlueDisable("Press me!"))
        ButtonView(ButtonState.BlueOutline("Press me!") {})
        ButtonView(ButtonState.BlueOutlineDisabled("Press me!"))
        ButtonView(ButtonState.BlueTextOnly("Press me!") {})
        ButtonView(ButtonState.BlueTextOnlyDisabled("Press me!"))
    }
}